File: qlik.dev/tutorials/platform-operations/sdk-python/tenant_configure.py
"""
Implementation of the 'Configure a tenant' tutorial from https://qlik.dev/tutorials/configure-a-tenant

For a detailed overview of the supported arguments execute:

    python tenant_configure.py --help
"""
import argparse
import json
import logging

from argparse_logging import add_log_level_argument
from qlik_sdk import AssignmentCreate, SpaceCreate

import constants
import qlik_sdk_helper
from jwt_auth import JwtAuth, JwtIdpConfig

logger = logging.getLogger(__name__)


def get_tenant_id(sdk_client):
    user = sdk_client.users.get_me()

    logger.info(f"Retrieved tenant ID from tenant '{sdk_client.config.host}'.")
    return user["tenantId"]


def enable_auto_group_creation(sdk_client):
    sdk_client.rest(
        path="/api/v1/groups/settings",
        method="PATCH",
        data=[{
            "op": "replace",
            "path": "/autoCreateGroups",
            "value": True
        }])

    logger.info(f"Enabled group auto creation on tenant '{sdk_client.config.host}'.")


def enable_auto_license_assignment(sdk_client):
    sdk_client.rest(
        path="/api/v1/licenses/settings",
        method="PUT",
        data={
            "autoAssignProfessional": True,
            "autoAssignAnalyzer": True
        })

    logger.info(f"Enabled license auto assignment on tenant '{sdk_client.config.host}'.")


def configure_jwt_idp(sdk_client, jwt_idp_config):
    tenant_id = get_tenant_id(sdk_client)
    with open(jwt_idp_config.public_key_file_path, "r") as file:
        public_key = file.read().replace("\n", "")

    identity_provider = json.loads(sdk_client.rest(
        path="/api/v1/identity-providers",
        method="POST",
        data={
            "tenantIds": [tenant_id],
            "provider": "external",
            "protocol": "jwtAuth",
            "interactive": False,
            "active": True,
            "description": "IdP to handle deferred authentication.",
            "options": {
                "jwtLoginEnabled": True,
                "issuer": jwt_idp_config.issuer,
                "staticKeys": [
                    {
                        "kid": jwt_idp_config.key_id,
                        "pem": public_key
                    }
                ]
            }
        }).text)

    logger.info(
        f"Created JWT identity provider with ID '{identity_provider['id']}' in tenant '{sdk_client.config.host}'.")


def create_shared_space(sdk_client):
    space = sdk_client.spaces.create(SpaceCreate(
        name=constants.SPACE_SHARED_DEV,
        type="shared"))

    logger.info(f"Created the shared space '{space.name}' with ID '{space.id}' in tenant '{sdk_client.config.host}'.")
    return space


def create_managed_space(sdk_client):
    space = sdk_client.spaces.create(SpaceCreate(
        name=constants.SPACE_MANAGED_PROD,
        type="managed"))

    logger.info(f"Created the managed space '{space.name}' with ID '{space.id}' in tenant '{sdk_client.config.host}'.")
    return space


def create_group(sdk_client, group_name, jwt_idp_config):
    jwt_auth = JwtAuth(sdk_client.config.host,
                       jwt_idp_config,
                       subject="temp_user",
                       name="temp_user",
                       email="temp_user@jwt.io",
                       groups=[group_name])
    user = json.loads(jwt_auth.rest(
        path="/api/v1/users/me",
        method="GET").text)

    logger.info(
        f"Created a JWT authentication session for a user in group '{group_name}' in tenant '{sdk_client.config.host}'.")

    # Lookup the newly created group to get the ID
    groups = json.loads(sdk_client.rest(
        path="/api/v1/groups",
        method="GET").text)["data"]

    group_id = None
    for group in groups:
        if group["name"] == group_name:
            group_id = group["id"]
            break

    if not group_id:
        logger.error(f"The group {group_name} could not be found in tenant '{sdk_client.config.host}'.")

    logger.info(f"Created group '{group_name}' with ID '{group_id}' in '{sdk_client.config.host}'.")

    # Delete the temporary user, it's not needed
    sdk_client.rest(
        path=f"/api/v1/users/{user['id']}",
        method="DELETE")

    logger.info(f"Deleted temporary user with ID '{user['id']}' from '{sdk_client.config.host}'.")

    return group_id


def assign_to_space(sdk_client, space, group_id, roles):
    space.create_assignment(AssignmentCreate(
        type="group",
        assigneeId=group_id,
        roles=roles
    ))

    logger.info(
        f"Assigned the group with ID '{group_id}' to the space with ID '{space.id}' with the roles '{roles}' in tenant '{sdk_client.config.host}'.")


def run(target_tenant_sdk_client, jwt_idp_config):
    enable_auto_group_creation(target_tenant_sdk_client)
    enable_auto_license_assignment(target_tenant_sdk_client)

    configure_jwt_idp(target_tenant_sdk_client, jwt_idp_config)

    dev_space = create_shared_space(target_tenant_sdk_client)
    prod_space = create_managed_space(target_tenant_sdk_client)

    analytics_consumer_group_id = create_group(target_tenant_sdk_client, constants.GROUP_ANALYTICS_CONSUMER,
                                               jwt_idp_config)
    assign_to_space(target_tenant_sdk_client, prod_space, analytics_consumer_group_id, ["consumer"])

    logger.info(f"The tenant '{target_tenant_sdk_client.config.host}' has been configured.")
    return dev_space.id, prod_space.id


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    add_log_level_argument(parser)
    parser.add_argument("--client-id", required=True, help="The OAuth client ID.")
    parser.add_argument("--client-secret", required=True, help="The OAuth client secret.")
    parser.add_argument("--target-tenant-hostname", required=True,
                        help="The hostname of the target tenant to configure, for example: tenant.region.qlikcloud.com")

    jwt_group = parser.add_argument_group("Target JWT IdP Configuration")
    jwt_group.add_argument("--jwt-issuer", required=False, help="The 'issuer' field to use in the JWT.")
    jwt_group.add_argument("--jwt-key-id", required=False, help="The 'kid' field to use in the JWT.")
    jwt_group.add_argument("--jwt-private-key", required=False, help="The path to the local private key file.")
    jwt_group.add_argument("--jwt-public-key", required=False, help="The path to the local public key file.")

    args = parser.parse_args()
    logging.basicConfig(level=args.log_level)

    jwt_idp_config = JwtIdpConfig(args.jwt_issuer, args.jwt_key_id, args.jwt_private_key, args.jwt_public_key)
    if not jwt_idp_config.validate():
        parser.print_help()
        exit(1)

    target_tenant_sdk_client = qlik_sdk_helper.create_sdk_client(args.client_id, args.client_secret,
                                                                 args.target_tenant_hostname)

    run(target_tenant_sdk_client, jwt_idp_config)

File: qlik.dev/tutorials/platform-operations/sdk-typescript/tenant_configure.js
import { getMyUser } from '@qlik/api/users';
import { patchGroupsSettings, patchGroup, createGroup as createGroupApi, getGroups } from '@qlik/api/groups';
import { getLicenseSettings, updateLicenseSettings } from '@qlik/api/licenses';
import { createSpace, getSpaces, getSpace, createSpaceAssignment } from '@qlik/api/spaces';
import { SPACE_SHARED_DEV, SPACE_MANAGED_PROD, GROUP_ANALYTICS_CONSUMER, ROLE_ANALYTICS_ADMIN } from './constants.js';

async function getTenantId(hostConfig) {
  const { data: me } = await getMyUser({ hostConfig });
  console.info(`Retrieved tenant ID '${me.tenantId}' from tenant '${hostConfig.host}'.`);
  return me.tenantId;
}

async function enableAutoGroupCreation(hostConfig) {
  await patchGroupsSettings([
    { op: 'replace', path: '/autoCreateGroups', value: true },
    { op: 'replace', path: '/syncIdpGroups', value: true },
  ], { hostConfig });
  console.info(`Enabled group auto creation on tenant '${hostConfig.host}'.`);
}

async function enableAutoLicenseAssignment(hostConfig) {
  const { data: originalSettings } = await getLicenseSettings({ hostConfig });
  await updateLicenseSettings(
    {
      ...originalSettings,
      autoAssignAnalyzer: true,
      autoAssignProfessional: false,
    },
    { hostConfig },
  );
  console.info(`Enabled license auto assignment on tenant '${hostConfig.host}'.`);
}

async function createOrGetSpace(hostConfig, spaceType, spaceName) {
  let space = false;
  try {
    const { data: created } = await createSpace({ name: spaceName, type: spaceType }, { hostConfig });
    space = created;
    console.info(`Created the ${spaceType} space '${space.name}' with ID '${space.id}' in tenant '${hostConfig.host}'.`);
  } catch (error) {
    console.info(`Failed to create the ${spaceType} space '${spaceName}', will try to get the space if it exists`);
  }
  if (!space) {
    try {
      const { data: spacesResponse } = await getSpaces({ name: spaceName, type: spaceType }, { hostConfig });
      const { data: found } = await getSpace(spacesResponse.data[0].id, { hostConfig });
      space = found;
      console.info(`Found existing ${spaceType} space '${spaceName}'`);
    } catch (error) {
      console.info(`Failed to get the ${spaceType} space '${spaceName}'`);
    }
  }
  return space;
}

async function createSharedSpace(hostConfig) {
  return createOrGetSpace(hostConfig, 'shared', SPACE_SHARED_DEV);
}

async function createManagedSpace(hostConfig) {
  return createOrGetSpace(hostConfig, 'managed', SPACE_MANAGED_PROD);
}

async function createOrGetGroup(hostConfig, groupName) {
  const adminConfig = { ...hostConfig, scope: 'admin_classic user_default' };
  let groupId;
  try {
    const { data: group } = await createGroupApi({ name: groupName }, { hostConfig: adminConfig });
    groupId = group.id;
    console.info(`Created group '${groupName}' with ID '${groupId}' in '${hostConfig.host}'.`);
  } catch (error) {
    console.info(`Failed to create group '${groupName}', will try to get it if it exists`);
  }
  if (!groupId) {
    try {
      const { data: groupsResponse } = await getGroups({ filter: `name eq "${groupName}"` }, { hostConfig });
      const existing = (groupsResponse?.data ?? []).find(({ name }) => name === groupName);
      if (!existing) {
        console.error(`Group '${groupName}' could not be found in tenant '${hostConfig.host}'.`);
        process.exit(1);
      }
      groupId = existing.id;
      console.info(`Found existing group '${groupName}' with ID '${groupId}' in '${hostConfig.host}'.`);
    } catch (error) {
      console.error(`Failed to get group '${groupName}' from '${hostConfig.host}'.`);
      process.exit(1);
    }
  }
  return groupId;
}

async function assignRolesToGroup(hostConfig, groupId, roles) {
  await patchGroup(
    groupId,
    [{ op: 'replace', path: '/assignedRoles', value: roles.map((name) => ({ name })) }],
    { hostConfig },
  );
  console.info(`Assigned roles '${roles}' to group '${groupId}' in tenant '${hostConfig.host}'.`);
}

async function assignToSpace(hostConfig, space, groupId, roles) {
  try {
    await createSpaceAssignment(
      space.id,
      { type: 'group', assigneeId: groupId, roles },
      { hostConfig },
    );
    console.info(`Assigned group '${groupId}' to space '${space.id}' with roles '${roles}' in tenant '${hostConfig.host}'.`);
  } catch (error) {
    if (error.data?.detail === 'assignment of assignee already exists in space') {
      console.info('Assignment already exists in space');
    } else {
      console.warn('Failed to create assignment');
      console.warn(error);
    }
  }
}

export const runTenantConfigure = async (hostConfig) => {
  const tenantId = await getTenantId(hostConfig);
  console.info(`Configuring tenantId: ${tenantId}`);
  await enableAutoGroupCreation(hostConfig);
  await enableAutoLicenseAssignment(hostConfig);

  const devSpace = await createSharedSpace(hostConfig);
  const prodSpace = await createManagedSpace(hostConfig);
  const analyticsConsumerGroupId = await createOrGetGroup(hostConfig, GROUP_ANALYTICS_CONSUMER);
  await assignRolesToGroup(hostConfig, analyticsConsumerGroupId, [ROLE_ANALYTICS_ADMIN]);
  await assignToSpace(hostConfig, prodSpace, analyticsConsumerGroupId, ['consumer']);

  console.info(`The tenant '${hostConfig.host}' has been configured.`);
  return { devSpaceId: devSpace.id, prodSpaceId: prodSpace.id, analyticsConsumerGroupId };
};
